server/src/main/java/org/candlepin/async/JobManager.java

```diff
@@ -139,13 +147,54 @@ public JobManager(
         final PrincipalProvider principalProvider,
         final CandlepinRequestScope scope,
         final Injector injector) {
+
         this.jobCurator = Objects.requireNonNull(jobCurator);
         this.dispatcher = Objects.requireNonNull(dispatcher);
         this.candlepinRequestScope = Objects.requireNonNull(scope);
         this.principalProvider = Objects.requireNonNull(principalProvider);
         this.injector = Objects.requireNonNull(injector);
     }
 
+    /**
+     * Calculates a hash for the unique job constraints of a given job. If the provided map of
+     * constraints is null or empty, this method returns null.
+     *
+     * @param constraints
+     *  A map of unique job constraints for which to calculate a hash
+     *
+     * @return
+     *  a hash of the provided job constraints, or null if no constraints were provided
+     */
+    private String calculateUniqueJobConstraintHash(Map<String, String> constraints) {
```

Comment:
I'm not very familiar with the jobs framework (the new, or the old), so the following scenario might never happen, but here's what I want to ask:
- If you have job1 with a constraint on owner_id=my_onwer currently running, you don't want another job of this type with the same constraint running concurrently.
- And job2 is queued with two constraints: owner_id=my_owner, consumer_id=my_consumer.
- These 2 jobs are going to have different hashes from what I understand, so we'll end up running them concurrently.

Does that not violate the logic behind job1's constraint (because we end up running 2 jobs against the same owner_id, even though one of them is more specific in its restriction)? So the main question is: there a possibility of the same Job type being run with different sets/subsets/supersets of constraints?

Reply:
That's a good question and part of what @mstead and I discussed Tuesday, and the focus of the query I brought up post-scrum yesterday. The answer, however, is not as clear since it's net-new functionality.

The intent behind converting it to a hash is to allow for arbitrary constraints per-job, while making the constraint restriction feel intuitive, regardless of the job order. This brought up a number of questions regarding the restriction, like the one you've posted above. Namely:

- Should *any* constraint match conflict?
- Should matching all of the constraints on the inbound job, but not on the existing job conflict (non-commutative matching)?
- Should order of the constraints matter?

For this implementation, we've opted with the answers, no, no and no. But as to whether or not that's correct or best... I'm not so sure. Just in responding to this post I've come up with a scenario where we may want to revert to what I had on Tuesday and following what you've said above. I.e.:

Imagine we have a job "refresh_owner_product" which updates an owner's product definition from upstream. Generally it would probably have constraints owner_id and product_id, as we probably wouldn't mind two different products being refreshed in parallel. However, two potential issues come up here:
- What if we wanted to allow the job to refresh *all* products for the org by not specifying a product. In such a case, we'd want the job to be constrained by the only constraint we passed in; which would be only the owner_id
- What if we decided that refreshing two products in parallel causes some kind of locking problem (maybe updating pools or something), but still want to be sure no one else refreshes the product at the same time (deduplication issues, perhaps). In this case, we'd want to make sure any constraint collisions blocks.

I'm less sold on the second scenario, but the first is definitely something I can see being a problem going forward. There's also a minor issue when it comes to debugging with the hashing method in that we 'lose' the inputs, so it'll be impossible for a human to know why two hashes collide without hands-on debugging.

Maybe this is something we should have a larger discussion on before continuing.